README.rst
=========
pyzerproc
=========


.. image:: https://img.shields.io/pypi/v/pyzerproc.svg
        :target: https://pypi.python.org/pypi/pyzerproc

.. image:: https://img.shields.io/travis/emlove/pyzerproc.svg
        :target: https://travis-ci.org/emlove/pyzerproc

.. image:: https://coveralls.io/repos/emlove/pyzerproc/badge.svg
        :target: https://coveralls.io/r/emlove/pyzerproc


Library to control Zerproc Bluetooth LED smart string lights

* Free software: Apache Software License 2.0


Features
--------

* Discover nearby devices
* Turn lights on and off
* Set light color
* Get light status


Command line usage
------------------
pyzerproc ships with a command line tool that exposes the features of the library.

.. code-block:: console

    $ pyzerproc discover
    INFO:pyzerproc.discovery:Starting scan for local devices
    INFO:pyzerproc.discovery:Discovered AA:BB:CC:00:11:22: LEDBlue-CC001122
    INFO:pyzerproc.discovery:Discovered AA:BB:CC:33:44:55: LEDBlue-CC334455
    INFO:pyzerproc.discovery:Scan complete
    AA:BB:CC:00:11:22
    AA:BB:CC:33:44:55

    $ pyzerproc turn-on AA:BB:CC:00:11:22
    INFO:pyzerproc.light:Connecting to AA:BB:CC:00:11:22
    INFO:pyzerproc.light:Turning on AA:BB:CC:00:11:22

    $ pyzerproc turn-off AA:BB:CC:00:11:22
    INFO:pyzerproc.light:Connecting to AA:BB:CC:00:11:22
    INFO:pyzerproc.light:Turning off AA:BB:CC:00:11:22

    $ pyzerproc set-color AA:BB:CC:00:11:22 ff0000
    INFO:pyzerproc.light:Connecting to AA:BB:CC:00:11:22
    INFO:pyzerproc.light:Changing color of AA:BB:CC:00:11:22 to #ff0000

    $ pyzerproc set-color AA:BB:CC:00:11:22 00ff00
    INFO:pyzerproc.light:Connecting to AA:BB:CC:00:11:22
    INFO:pyzerproc.light:Changing color of AA:BB:CC:00:11:22 to #00ff00

    $ pyzerproc is-on AA:BB:CC:00:11:22
    INFO:pyzerproc.light:Connecting to AA:BB:CC:00:11:22
    INFO:pyzerproc.light:Got state of AA:BB:CC:00:11:22: <LightState is_on='True' color='(255, 0, 0)'>
    True

    $ pyzerproc get-color AA:BB:CC:00:11:22
    INFO:pyzerproc.light:Connecting to AA:BB:CC:00:11:22
    INFO:pyzerproc.light:Got state of AA:BB:CC:00:11:22: <LightState is_on='True' color='(255, 0, 0)'>
    ff0000


Usage
-----

Discover nearby devices

.. code-block:: python

    import pyzerproc

    lights = pyzerproc.discover(timeout=30)

    for light in lights:
        print("Address: {} Name: {}".format(light.address, light.name))


Turn a light on and off

.. code-block:: python

    import pyzerproc
    import time

    address = "AA:BB:CC:00:11:22"

    light = pyzerproc.Light(address)

    try:
        light.connect(auto_reconnect=True)
        light.turn_on()

        time.sleep(5)

        light.turn_off()
    finally:
        light.disconnect()


Change the light color

.. code-block:: python

    import pyzerproc
    import time

    address = "AA:BB:CC:00:11:22"

    light = pyzerproc.Light(address)

    try:
        light.connect()

        while True:
            light.set_color(255, 0, 0) # Red
            time.sleep(1)
            light.set_color(0, 255, 0) # Green
            time.sleep(1)
    finally:
        light.disconnect()


Get the light state

.. code-block:: python

    import pyzerproc
    import time

    address = "AA:BB:CC:00:11:22"

    light = pyzerproc.Light(address)

    try:
        light.connect()

        state = light.get_state()

        if state.is_on:
            print(state.color)
        else:
            print("Off")
    finally:
        light.disconnect()


Changelog
---------
Next
~~~~
- Expose exception objects
- Expose light address and name on discovery

0.1.1 (2020-05-08)
~~~~~~~~~~~~~~~~~~
- Expose auto reconnect

0.1.0 (2020-05-07)
~~~~~~~~~~~~~~~~~~
- Discover nearby devices

0.0.2 (2020-05-05)
~~~~~~~~~~~~~~~~~~
- Get the current light state

0.0.1 (2020-05-04)
~~~~~~~~~~~~~~~~~~
- Initial release


Credits
-------

- Thanks to `Uri Shaked`_ for an incredible guide to `Reverse Engineering a Bluetooth Lightbulb`_.

- This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _`Uri Shaked`: https://medium.com/@urish
.. _`Reverse Engineering a Bluetooth Lightbulb`: https://medium.com/@urish/reverse-engineering-a-bluetooth-lightbulb-56580fcb7546
.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
=======
History
=======

0.0.1 (2020-05-03)
------------------

* First release on PyPI.


tests/conftest.py
import sys
import pytest

@pytest.fixture
def adapter(mocker):
    return mocker.patch('pygatt.GATTToolBackend')()

@pytest.fixture
def device(mocker, adapter):
    device = mocker.Mock()
    adapter.connect.return_value = device
    return device


tests/test_light.py
#!/usr/bin/env python
import queue
import pytest

from pyzerproc import Light


def test_connect_disconnect(adapter, device):
    """Test the CLI."""
    light = Light("00:11:22")
    light.connect()

    adapter.start.assert_called_with(reset_on_start=False)
    adapter.connect.assert_called_with("00:11:22")
    device.subscribe.assert_called_with(
        "0000ffe4-0000-1000-8000-00805f9b34fb", callback=light._handle_data)

    light.disconnect()

    adapter.stop.assert_called_once()

    # Duplicate disconnect shouldn't call stop again
    light.disconnect()

    adapter.stop.assert_called_once()


def test_turn_on_not_connected(device):
    """Test the CLI."""
    light = Light("00:11:22")

    with pytest.raises(RuntimeError):
        light.turn_on()


def test_turn_on(device):
    """Test the CLI."""
    light = Light("00:11:22")
    light.connect()

    light.turn_on()

    device.char_write.assert_called_with(
        '0000ffe9-0000-1000-8000-00805f9b34fb', b'\xCC\x23\x33')


def test_turn_off(device):
    """Test the CLI."""
    light = Light("00:11:22")
    light.connect()

    light.turn_off()

    device.char_write.assert_called_with(
        '0000ffe9-0000-1000-8000-00805f9b34fb', b'\xCC\x24\x33')


def test_set_color(device):
    """Test the CLI."""
    light = Light("00:11:22")
    light.connect()

    light.set_color(0, 0, 0)
    device.char_write.assert_called_with(
        '0000ffe9-0000-1000-8000-00805f9b34fb',
        b'\x56\x00\x00\x00\x00\xF0\xAA')

    light.set_color(255, 255, 255)
    device.char_write.assert_called_with(
        '0000ffe9-0000-1000-8000-00805f9b34fb',
        b'\x56\x1F\x1F\x1F\x00\xF0\xAA')

    light.set_color(64, 128, 192)
    device.char_write.assert_called_with(
        '0000ffe9-0000-1000-8000-00805f9b34fb',
        b'\x56\x07\x0F\x17\x00\xF0\xAA')

    with pytest.raises(ValueError):
        light.set_color(999, 999, 999)


def test_get_state(device, mocker):
    """Test the CLI."""
    light = Light("00:11:22")
    light.connect()

    def send_response(*args, **kwargs):
        """Simulate a response from the light"""
        light._handle_data(
            63, b'\x66\xe3\x24\x16\x24\x01\xff\x00\x00\x00\x01\x99')

    device.char_write.side_effect = send_response

    state = light.get_state()
    assert state.is_on is False
    assert state.color == (255, 0, 0)
    device.char_write.assert_called_with(
        '0000ffe9-0000-1000-8000-00805f9b34fb',  b'\xEF\x01\x77')
    assert state.__repr__() == "<LightState is_on='False' color='(255, 0, 0)'>"

    # Ensure duplicate responses are handled
    def send_response(*args, **kwargs):
        """Simulate a response from the light"""
        light._handle_data(
            63, b'\x66\xe3\x23\x16\x24\x01\x10\x05\x1C\x00\x01\x99')
        light._handle_data(
            63, b'\x66\xe3\x23\x16\x24\x01\x10\x05\x1C\x00\x01\x99')

    device.char_write.side_effect = send_response

    state = light.get_state()
    assert state.is_on is True
    assert state.color == (131, 41, 230)
    device.char_write.assert_called_with(
        '0000ffe9-0000-1000-8000-00805f9b34fb',  b'\xEF\x01\x77')

    # Ensure leftover values are discarded before querying
    def send_response(*args, **kwargs):
        """Simulate a response from the light"""
        light._handle_data(
            63, b'\x66\xe3\x00\x16\x24\x01\xFF\xFF\xFF\x00\x01\x99')

    device.char_write.side_effect = send_response

    state = light.get_state()
    assert state.is_on is None
    assert state.color == (255, 255, 255)
    device.char_write.assert_called_with(
        '0000ffe9-0000-1000-8000-00805f9b34fb',  b'\xEF\x01\x77')

    # Test response timeout
    device.char_write.side_effect = None
    light.notification_queue = mocker.MagicMock()

    def get_queue(*args, **kwargs):
        """Simulate a queue timeout"""
        raise queue.Empty()

    light.notification_queue.get.side_effect = get_queue

    with pytest.raises(TimeoutError):
        state = light.get_state()


tests/test_pyzerproc.py
#!/usr/bin/env python

"""Tests for `pyzerproc` package."""

import pytest

from click.testing import CliRunner

from pyzerproc import pyzerproc
from pyzerproc import cli


@pytest.fixture
def response():
    """Sample pytest fixture.

    See more at: http://doc.pytest.org/en/latest/fixture.html
    """
    # import requests
    # return requests.get('https://github.com/audreyr/cookiecutter-pypackage')


def test_content(response):
    """Sample pytest test function with the pytest fixture as an argument."""
    # from bs4 import BeautifulSoup
    # assert 'GitHub' in BeautifulSoup(response.content).title.string


def test_command_line_interface():
    """Test the CLI."""
    runner = CliRunner()
    result = runner.invoke(cli.main)
    assert result.exit_code == 0
    assert 'pyzerproc.cli.main' in result.output
    help_result = runner.invoke(cli.main, ['--help'])
    assert help_result.exit_code == 0
    assert '--help  Show this message and exit.' in help_result.output


pyzerproc/cli.py
"""Console script for pyzerproc."""
import sys
from binascii import hexlify
import click
import logging

import pyzerproc


@click.group()
@click.option('-v', '--verbose', count=True,
              help="Pass once to enable pyzerproc debug logging. Pass twice "
                   "to also enable pygatt debug logging.")
def main(verbose):
    """Console script for pyzerproc."""
    logging.basicConfig()
    logging.getLogger('pyzerproc').setLevel(logging.INFO)
    if verbose >= 1:
        logging.getLogger('pyzerproc').setLevel(logging.DEBUG)
    if verbose >= 2:
        logging.getLogger('pygatt').setLevel(logging.DEBUG)


@main.command()
@click.argument('address')
def turn_on(address):
    """Turn on the light with the given MAC address"""
    light = pyzerproc.Light(address)

    try:
        light.connect()
        light.turn_on()
    finally:
        light.disconnect()
    return 0


@main.command()
@click.argument('address')
def turn_off(address):
    """Turn off the light with the given MAC address"""
    light = pyzerproc.Light(address)

    try:
        light.connect()
        light.turn_off()
    finally:
        light.disconnect()
    return 0


@main.command()
@click.argument('address')
@click.argument('color')
def set_color(address, color):
    """Set the light with the given MAC address to an RRGGBB hex color"""
    light = pyzerproc.Light(address)

    r, g, b = tuple(int(color[i:i+2], 16) for i in (0, 2, 4))

    try:
        light.connect()
        light.set_color(r, g, b)
    finally:
        light.disconnect()
    return 0


@main.command()
@click.argument('address')
def is_on(address):
    """Get the current on/off status of the light"""
    light = pyzerproc.Light(address)

    try:
        light.connect()
        state = light.get_state()
        click.echo(state.is_on)
    finally:
        light.disconnect()
    return 0


@main.command()
@click.argument('address')
def get_color(address):
    """Get the current color of the light"""
    light = pyzerproc.Light(address)

    try:
        light.connect()
        state = light.get_state()
        click.echo(hexlify(bytes(state.color)))
    finally:
        light.disconnect()
    return 0


if __name__ == "__main__":
    sys.exit(main())  # pragma: no cover


pyzerproc/discovery.py
"""Device discovery code"""
import logging

import pygatt

from .light import Light

_LOGGER = logging.getLogger(__name__)


def discover(timeout=10):
    """Returns nearby discovered lights."""
    _LOGGER.info("Starting scan for local devices")

    adapter = pygatt.GATTToolBackend()
    adapter.start(reset_on_start=False)

    lights = []
    try:
        for device in adapter.scan(timeout=timeout):
            # Improvements welcome
            if device['name'] and device['name'].startswith('LEDBlue-'):
                _LOGGER.info(
                    "Discovered %s: %s", device['address'], device['name'])
                lights.append(
                    Light(device['address'], device['name'].strip()))
    finally:
        adapter.stop()

    _LOGGER.info("Scan complete")
    return lights


pyzerproc/light.py
"""Device class"""
from binascii import hexlify
import logging
import queue

import pygatt

_LOGGER = logging.getLogger(__name__)

CHARACTERISTIC_COMMAND_WRITE = "0000ffe9-0000-1000-8000-00805f9b34fb"
CHARACTERISTIC_NOTIFY_VALUE = "0000ffe4-0000-1000-8000-00805f9b34fb"

NOTIFICATION_RESPONSE_TIMEOUT = 5


class Light():
    """Represents one connected light"""

    def __init__(self, address, name=None):
        self.address = address
        self.name = name
        self.adapter = None
        self.device = None
        self.notification_queue = queue.Queue(maxsize=1)

    def connect(self, auto_reconnect=False):
        """Connect to this light"""
        _LOGGER.info("Connecting to %s", self.address)

        self.adapter = pygatt.GATTToolBackend()
        self.adapter.start(reset_on_start=False)
        self.device = self.adapter.connect(
            self.address, auto_reconnect=auto_reconnect)

        _LOGGER.debug("Connected to %s", self.address)

        self.device.subscribe(CHARACTERISTIC_NOTIFY_VALUE,
                              callback=self._handle_data)

    def disconnect(self):
        """Connect to this light"""
        if self.adapter:
            self.adapter.stop()
            self.adapter = None
            self.device = None

    def turn_on(self):
        """Turn on the light"""
        _LOGGER.info("Turning on %s", self.address)
        self._write(CHARACTERISTIC_COMMAND_WRITE, b'\xCC\x23\x33')
        _LOGGER.debug("Turned on %s", self.address)

    def turn_off(self):
        """Turn off the light"""
        _LOGGER.info("Turning off %s", self.address)
        self._write(CHARACTERISTIC_COMMAND_WRITE, b'\xCC\x24\x33')
        _LOGGER.debug("Turned off %s", self.address)

    def set_color(self, r, g, b):
        """Set the color of the light

        Accepts red, green, and blue values from 0-255
        """
        for value in (r, g, b):
            if not 0 <= value <= 255:
                raise ValueError(
                    "Value {} is outside the valid range of 0-255")

        _LOGGER.info("Changing color of %s to #%02x%02x%02x",
                     self.address, r, g, b)

        # Normalize to 0-31, the supported range of these lights
        r = int(r * 31 / 255)
        g = int(g * 31 / 255)
        b = int(b * 31 / 255)
        color_string = "{:c}{:c}{:c}".format(r, g, b).encode()

        value = b'\x56' + color_string + b'\x00\xF0\xAA'
        self._write(CHARACTERISTIC_COMMAND_WRITE, value)
        _LOGGER.debug("Changed color of %s", self.address)

    def _handle_data(self, handle, value):
        """Handle an incoming notification message."""
        _LOGGER.debug("Got handle '%s' and value %s", handle, hexlify(value))
        try:
            self.notification_queue.put_nowait(value)
        except queue.Full:
            _LOGGER.debug("Discarding duplicate response", exc_info=True)

    def get_state(self):
        """Get the current state of the light"""
        # Clear the queue if a value is somehow left over
        try:
            self.notification_queue.get_nowait()
        except queue.Empty:
            pass

        self._write(CHARACTERISTIC_COMMAND_WRITE, b'\xEF\x01\x77')

        try:
            response = self.notification_queue.get(
                timeout=NOTIFICATION_RESPONSE_TIMEOUT)
        except queue.Empty as ex:
            raise TimeoutError("Timeout waiting for response") from ex

        on_off_value = int(response[2])

        r = int(response[6])
        g = int(response[7])
        b = int(response[8])

        if on_off_value == 0x23:
            is_on = True
        elif on_off_value == 0x24:
            is_on = False
        else:
            is_on = None

        # Normalize and clamp from 0-31, to 0-255
        r = int(min(r * 255 / 31, 255))
        g = int(min(g * 255 / 31, 255))
        b = int(min(b * 255 / 31, 255))

        state = LightState(is_on, (r, g, b))

        _LOGGER.info("Got state of %s: %s", self.address, state)

        return state

    def _write(self, uuid, value):
        """Internal method to write to the device"""
        if not self.device:
            raise RuntimeError(
                "Light {} is not connected".format(self.address))

        _LOGGER.debug("Writing 0x%s to characteristic %s", value.hex(), uuid)
        self.device.char_write(uuid, value)
        _LOGGER.debug("Wrote 0x%s to characteristic %s", value.hex(), uuid)


class LightState():
    """Represents the current state of the light"""
    __slots__ = 'is_on', 'color',

    def __init__(self, is_on, color):
        """Create the state object"""
        self.is_on = is_on
        self.color = color

    def __repr__(self):
        """Return a string representation of the state object"""
        return "<LightState is_on='{}' color='{}'>".format(
            self.is_on, self.color)

    def __eq__(self, other):
        """Check for equality."""
        return self.is_on == other.is_on and self.color == other.color
